velox/serializers/RowSerializer.h

```diff
@@ -99,20 +141,69 @@ class RowSerializer : public IterativeVectorSerializer {
   }
 
   size_t maxSerializedSize() const override {
-    size_t totalSize = 0;
-    for (const auto& buffer : buffers_) {
-      totalSize += buffer->size();
+    const auto size = uncompressedSize();
+    if (!needCompression()) {
+      return detail::RowHeader::size() + size;
     }
-    return totalSize;
+    VELOX_CHECK_LE(
+        size,
+        codec_->maxUncompressedLength(),
+        "UncompressedSize exceeds limit");
+    return detail::RowHeader::size() + codec_->maxCompressedLength(size);
   }
 
+  // The serialization format is | uncompressedSize | compressedSize |
+  // compressed | data.
   void flush(OutputStream* stream) override {
-    for (const auto& buffer : buffers_) {
-      stream->write(buffer->template asMutable<char>(), buffer->size());
+    constexpr int32_t kMaxCompressionAttemptsToSkip = 30;
+    const auto size = uncompressedSize();
+    if (!needCompression()) {
+      flushUncompressed(size, stream);
+    } else if (numCompressionToSkip_ > 0) {
+      flushUncompressed(size, stream);
+      stats_.compressionSkippedBytes += size;
+      --numCompressionToSkip_;
+      ++stats_.numCompressionSkipped;
+    } else {
+      // Compress the buffer if satisfied condition.
+      const auto toCompress = toIOBuf(buffers_);
+      const auto compressedBuffer = codec_->compress(toCompress.get());
+      const int32_t compressedSize = compressedBuffer->length();
+      stats_.compressionInputBytes += size;
+      stats_.compressedBytes += compressedSize;
+      if (compressedSize > options_.minCompressionRatio * size) {
+        // Skip this compression.
+        numCompressionToSkip_ = std::min<int64_t>(
+            kMaxCompressionAttemptsToSkip, 1 + stats_.numCompressionSkipped);
+        flushUncompressed(size, stream);
+      } else {
+        // Do the compression.
+        detail::RowHeader header = {size, compressedSize, true};
+        header.write(stream);
+        for (auto range : *compressedBuffer) {
+          stream->write(
+              reinterpret_cast<const char*>(range.data()), range.size());
+        }
+      }
     }
+
     buffers_.clear();
   }
 
+  std::unordered_map<std::string, RuntimeCounter> runtimeStats() override {
```

Comment:
Shall move the non-trivial implementation to cc? thanks!

Reply:
The class RowSerializer has template row::CompactRow and row::UnsafeRow, so all the implements exists in header.